File: pipeline/lib/normalize.mjs
// Normalize one event's raw Smoothcomp pull into canonical adult no-gi matches.
// Handles BOTH bracket renderers: elimination (state.matches[]) and
// round-robin (state.rounds[n][]). Match objects share the same shape.
import { readFileSync, readdirSync, existsSync } from "node:fs";
import { join } from "node:path";

// "Men / Adult (age 15+) / Beginner / -60,0 kg" -> parts + derived flags.
function parseDivision(name) {
  const parts = String(name || "").split("/").map((s) => s.trim());
  const [gender = "", age = "", level = "", weight = ""] = parts;
  const blob = name.toLowerCase();
  let type = "standard";
  if (/super\s*fight|exhibition/.test(blob)) type = "exhibition";
  else if (/absolute|open\s*weight/.test(blob)) type = "absolute";
  return { gender, age, level, weight, type };
}

// Adult divisions across EN/PT/ES labels ("Adult", "Adults", "Adulto/a/s").
// Exclude youth (Boys/Girls/Kids/Juvenil/Infantil/Teen) and Masters/veterans —
// they never face adults, so cross-age ratings are not comparable.
const YOUTH_MASTER = /boy|girl|kid|juven|infan|mirim|teen|master|veteran|senior|master?s?\s*\d/i;
const isAdult = (age) => /adult/i.test(age) && !YOUTH_MASTER.test(age);
const DIRTY = new Set(["walkover", "disqualification", "dq", "wo"]);

// Pull the two competitors + winner + method out of one match object.
function readMatch(m, reg2user) {
  if (!m || typeof m !== "object" || m.isBye || m.isCancel) return null;
  const seats = m.seats || {};
  const L = seats.left, R = seats.right;
  if (!L || !R || !L.player || !R.player) return null;
  const idOf = (seat) => {
    const p = seat.player || {};
    // prefer the stable profile id from profile_link, else reg->user join,
    // else a name+country fallback key (resolved to a real id in build.mjs
    // when that person appears with a profile_link anywhere).
    const link = p.profile_link && String(p.profile_link).match(/profile\/(\d+)/);
    if (link) return Number(link[1]);
    const viaReg = reg2user.get(p.registration_id);
    if (viaReg) return viaReg;
    const nm = String(p.name || "").trim().toLowerCase().replace(/\s+/g, " ");
    return nm ? `n:${nm}|${(p.country || "").toLowerCase()}` : null;
  };
  const lid = idOf(L), rid = idOf(R);
  if (!lid || !rid || lid === rid) return null;
  let winner, loser;
  if (L.isWinner) { winner = lid; loser = rid; }
  else if (R.isWinner) { winner = rid; loser = lid; }
  else return null; // undecided
  const method = String(m.wonBy || "").toLowerCase() || "unknown";
  const names = new Map([[lid, L.player.name], [rid, R.player.name]]);
  const clubs = new Map([[lid, L.player.club], [rid, R.player.club]]);
  const countries = new Map([[lid, (L.player.country || "").toUpperCase()],
                             [rid, (R.player.country || "").toUpperCase()]]);
  return {
    match_id: m.id, round: Number(m.round || 0), method,
    is_clean: !DIRTY.has(method),
    score: (m.score && m.score.score) || "",
    winner, loser, names, clubs, countries,
  };
}

// Read participants.json -> registration_id -> user_id, plus a name book.
function loadParticipants(dir) {
  const reg2user = new Map();
  const book = new Map(); // user_id -> {name, club, country}
  const p = join(dir, "participants.json");
  if (!existsSync(p)) return { reg2user, book };
  const data = JSON.parse(readFileSync(p, "utf8"));
  for (const pt of data.participants || []) {
    for (const rg of pt.registrations || []) {
      if (rg.id && rg.user_id) reg2user.set(rg.id, rg.user_id);
      if (rg.user_id && !book.has(rg.user_id)) {
        book.set(rg.user_id, {
          name: `${rg.firstname || ""} ${rg.lastname || ""}`.trim() || `#${rg.user_id}`,
          // Smoothcomp participants use camelCase clubName; fall back to
          // team/affiliation, then the older snake_case just in case.
          club: rg.clubName || rg.teamName || rg.affiliationName || rg.club_name || rg.club || "",
          country: (rg.cn || rg.country || "").toUpperCase(),
        });
      }
    }
  }
  return { reg2user, book };
}

// Public: normalize one event directory -> { event, matches[], book }.
export function normalizeEvent(dir) {
  const event = JSON.parse(readFileSync(join(dir, "event.json"), "utf8"));
  const brackets = JSON.parse(readFileSync(join(dir, "brackets.json"), "utf8")).brackets || [];
  const div = new Map(); // bracket_id -> parsed division
  for (const b of brackets) div.set(b.bracket_id, { ...parseDivision(b.name), raw: b.name });
  const { reg2user, book } = loadParticipants(dir);

  const renderDir = join(dir, "render");
  const files = existsSync(renderDir) ? readdirSync(renderDir).filter((f) => f.endsWith(".json")) : [];
  const matches = [];
  for (const f of files) {
    const bracketId = Number(f.replace(".json", ""));
    const d = JSON.parse(readFileSync(join(renderDir, f), "utf8"));
    const st = d.state || {};
    // collect match objects from either renderer
    const raw = [];
    if (Array.isArray(st.matches)) raw.push(...st.matches);
    if (st.rounds && typeof st.rounds === "object")
      for (const arr of Object.values(st.rounds)) if (Array.isArray(arr)) raw.push(...arr);
    const d0 = div.get(bracketId) || {};
    if (!isAdult(d0.age) || d0.type === "exhibition") continue; // adult headline only
    for (const m of raw) {
      const parsed = readMatch(m, reg2user);
      if (!parsed) continue;
      // fold in athlete book entries from match seats: create for ids missing
      // from participants, and backfill a club when the participants entry had
      // none (the render seat carries the club/team string directly).
      for (const [id, nm] of parsed.names) {
        if (!id) continue;
        const seatClub = parsed.clubs.get(id) || "";
        if (!book.has(id)) {
          book.set(id, { name: nm || `#${id}`, club: seatClub, country: parsed.countries.get(id) || "" });
        } else if (seatClub && !book.get(id).club) {
          book.get(id).club = seatClub;
        }
      }
      matches.push({
        event_id: event.event_id, event_name: event.name, date: event.date,
        bracket_id: bracketId, division: d0.raw, division_type: d0.type,
        weight: d0.weight, gender: d0.gender, ruleset: event.ruleset || "nogi",
        match_id: parsed.match_id, round: parsed.round, method: parsed.method,
        is_clean: parsed.is_clean, score: parsed.score,
        winner: parsed.winner, loser: parsed.loser,
      });
    }
  }
  return { event, matches, book };
}

File: engine/normalize.py
#!/usr/bin/env python3
"""
Normalize raw Smoothcomp JSON (data/raw/*.json) into the canonical
data/matches.csv contract that engine/rate.py consumes.

Source files (all verified on disk, fetched by a prior ingest run this
session, event 30665 / ADCC US Open Austin):
  data/raw/brackets.json      -- 187 bracket/division listing records
  data/raw/participants.json  -- division roster: registration_id -> user_id
                                  (stable profile id), plus division name
  data/raw/render/<bracket_id>.json -- per-bracket match state (two schemas
                                  observed: {state:{rounds:{n:[...]}}} for
                                  single/double-elim brackets, and
                                  {state:{matches:[...] or {...}}} for
                                  round-robin/repechage brackets; both use
                                  the same per-match record shape)

KNOWN DEVIATIONS / EXCLUSIONS (logged, not silently dropped):
  1. Matches with wonBy in {"bye","double-bye"} are excluded -- one or both
     seats is a placeholder ("type":"bye","player":null), not a real
     contest between two competitors.
  2. Matches with wonBy == "double-dq" are excluded -- both sides show
     isWinner=false, no declared winner to score s in {0,1} against
     (model-spec.md S4 requires a winner; there isn't one for these).
  3. Matches referencing a registration_id absent from participants.json's
     roster are excluded -- these registrations are unapproved entries
     (seat player "approved":0 observed in every such case) not present in
     the roster snapshot this session pulled, so no stable profile_id
     (user_id) can be resolved for them. goal.md: "never invent ... a
     profile id" -- these are skipped rather than keyed by registration_id
     (registration_id is NOT the stable cross-event profile id).
  4. ruleset is hardcoded "nogi" for every match. Source data carries no
     per-division gi/nogi field (checked: no bracket name contains the
     token "gi" as a whole word, only the substring inside "Beginner").
     This event is on the adcc.smoothcomp.com tenant; ADCC is a federation
     that runs no-gi submission grappling exclusively (published ADCC
     rules, not re-fetched this session) -- inferred from tenant identity,
     not read from a per-match field. Flagged for re-verification the
     moment a mixed-ruleset event is added to the corpus.
  5. method vocabulary mapped from Smoothcomp's wonBy values to the
     model-spec.md S6 closed set: submission->submission, points->points,
     decision->decision, walkover->walkover (dirty), disqualification->dq
     (dirty). No "advantage" or "noshow"/"forfeit" wonBy value was observed
     in this event's data.
  6. belt_a/belt_b: ADCC divisions use skill levels (Beginner/Intermediate/
     Advanced), not BJJ belt colors. model-spec.md S3 explicitly allows
     "(or org equivalent)" for this field -- skill level is stored there.
  7. division name parsing is best-effort across three observed shapes
     (standard "Gender / AgeGroup / Skill / Weight", absolute "Gender's
     Absolute / AgeGroup / Skill" with no weight, and kids "Boys [11-12
     years] / Skill / Weight" where age is embedded in segment 0). All
     parsed fields are metadata-only (model-spec.md S3) -- they never feed
     the rating math, only display/audit.
"""
import csv
import json
import re
import sys
from pathlib import Path

RAW = Path("data/raw")
OUT = Path("data/matches.csv")
EVENT_ID = "30665"

DIRTY_METHOD_MAP = {
    "walkover": "walkover",
    "disqualification": "dq",
}
CLEAN_METHOD_MAP = {
    "submission": "submission",
    "points": "points",
    "decision": "decision",
}
EXCLUDE_WONBY = {"bye", "double-bye", "double-dq"}

WEIGHT_RE = re.compile(r"^[+-]\d+,\d+\s*kg$", re.I)


def iter_matches(container):
    return container.values() if isinstance(container, dict) else container


def load_registration_map(participants):
    reg_map = {}
    division_meta = {}
    for div in participants["participants"]:
        for reg in div["registrations"]:
            reg_map[reg["id"]] = reg["user_id"]
        division_meta[div["bracket_id"]] = {
            "division_id": div["id"],
            "name": div["name"],
        }
    return reg_map, division_meta


def parse_division_name(name):
    segs = [s.strip() for s in name.split("/")]
    gender = age_group = belt = weight_class = None
    division_type = "standard"

    if "Absolute" in segs[0]:
        division_type = "absolute"
        gender = segs[0].replace("'s Absolute", "").replace("Absolute", "").strip()
        if len(segs) >= 2:
            age_group = segs[1]
        if len(segs) >= 3:
            belt = segs[2]
    elif len(segs) >= 2 and re.match(r"^(Boys|Girls)\b", segs[0]):
        m = re.match(r"^(Boys|Girls)", segs[0])
        gender = m.group(1)
        age_group = segs[0]
        belt = segs[1] if len(segs) >= 2 else None
        if len(segs) >= 3 and WEIGHT_RE.match(segs[2]):
            weight_class = segs[2]
    elif len(segs) >= 4 and segs[0] in ("Men", "Women"):
        gender = segs[0]
        age_group = segs[1]
        belt = segs[2]
        weight_class = segs[3] if WEIGHT_RE.match(segs[3]) else None
    else:
        division_type = "exhibition"
        age_group = name

    return {
        "gender": gender or "",
        "age_group": age_group or "",
        "belt": belt or "",
        "weight_class": weight_class or "",
        "division_type": division_type,
    }


def main():
    brackets = json.loads((RAW / "brackets.json").read_text())["brackets"]
    participants = json.loads((RAW / "participants.json").read_text())
    reg_map, division_meta = load_registration_map(participants)

    bracket_render_ids = {b["bracket_id"] for b in brackets}

    rows = []
    excluded = {"bye": 0, "double_dq": 0, "unmapped_profile": 0}
    excluded_detail = []

    for bracket_id in sorted(bracket_render_ids):
        render_path = RAW / "render" / f"{bracket_id}.json"
        if not render_path.exists():
            excluded_detail.append(f"bracket {bracket_id}: no render file on disk")
            continue
        state = json.loads(render_path.read_text())["state"]
        all_matches = []
        if "rounds" in state:
            for _, ms in state["rounds"].items():
                all_matches.extend(iter_matches(ms))
        else:
            all_matches.extend(iter_matches(state["matches"]))

        meta = division_meta.get(bracket_id, {"division_id": "", "name": ""})
        parsed = parse_division_name(meta["name"]) if meta["name"] else {
            "gender": "", "age_group": "", "belt": "", "weight_class": "",
            "division_type": "",
        }

        for m in all_matches:
            won_by = m.get("wonBy")
            if won_by in EXCLUDE_WONBY:
                if won_by == "double-dq":
                    excluded["double_dq"] += 1
                else:
                    excluded["bye"] += 1
                continue

            left, right = m["seats"]["left"], m["seats"]["right"]
            if left.get("type") != "registration" or right.get("type") != "registration":
                excluded["bye"] += 1
                continue

            left_reg = left["player"]["registration_id"]
            right_reg = right["player"]["registration_id"]
            if left_reg not in reg_map or right_reg not in reg_map:
                excluded["unmapped_profile"] += 1
                excluded_detail.append(
                    f"match {m['id']} bracket {bracket_id}: "
                    f"left_reg={left_reg}({left_reg in reg_map}) "
                    f"right_reg={right_reg}({right_reg in reg_map})"
                )
                continue

            winner_side, loser_side = (left, right) if left.get("isWinner") else (right, left)
            winner_reg = winner_side["player"]["registration_id"]
            loser_reg = loser_side["player"]["registration_id"]

            if won_by in CLEAN_METHOD_MAP:
                method = CLEAN_METHOD_MAP[won_by]
                is_clean = "true"
            elif won_by in DIRTY_METHOD_MAP:
                method = DIRTY_METHOD_MAP[won_by]
                is_clean = "false"
            else:
                excluded_detail.append(f"match {m['id']}: unmapped wonBy={won_by!r}, excluded")
                continue

            rows.append({
                "match_id": m["id"],
                "date": m.get("estimated_starttime", ""),
                "event_id": EVENT_ID,
                "bracket_id": bracket_id,
                "round": m.get("round", ""),
                "ruleset": "nogi",
                "belt": parsed["belt"],
                "weight_class": parsed["weight_class"],
                "age_group": parsed["age_group"],
                "gender": parsed["gender"],
                "division_type": parsed["division_type"],
                "division_id": meta["division_id"],
                "winner_id": reg_map[winner_reg],
                "loser_id": reg_map[loser_reg],
                "winner_registration_id": winner_reg,
                "loser_registration_id": loser_reg,
                "method": method,
                "is_clean": is_clean,
            })

    fieldnames = [
        "match_id", "date", "event_id", "bracket_id", "round", "ruleset",
        "belt", "weight_class", "age_group", "gender", "division_type",
        "division_id", "winner_id", "loser_id", "winner_registration_id",
        "loser_registration_id", "method", "is_clean",
    ]
    with open(OUT, "w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=fieldnames)
        w.writeheader()
        for r in rows:
            w.writerow(r)

    print(f"wrote {len(rows)} rows to {OUT}")
    print(f"excluded: bye/no-opponent={excluded['bye']} "
          f"double-dq={excluded['double_dq']} "
          f"unmapped-profile={excluded['unmapped_profile']}")
    print(f"brackets processed: {len(bracket_render_ids)}")
    Path("data/normalize-exclusions-30665.txt").write_text("\n".join(excluded_detail) + "\n")
    print(f"exclusion detail written to data/normalize-exclusions-30665.txt "
          f"({len(excluded_detail)} lines)")


if __name__ == "__main__":
    main()
